### PR TITLE
Fix handling of User-Agent headers with bad utf-8.

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes error logs such as:

```
Failed to insert client IP ...  UnicodeEncodeError ... 'surrogates not allowed'
```

### CODE REVIEW

File: synapse/api/auth.py [hunk 1]
```diff
@@ -186,7 +186,7 @@ async def get_user_by_req(
             ip_addr = self.hs.get_ip_from_request(request)
             user_agent = request.requestHeaders.getRawHeaders(
                 b"User-Agent", default=[b""]
-            )[0].decode("ascii", "surrogateescape")
+            )[0].decode("ascii", "replace")
```
Comment (erikjohnston) [inline synapse/api/auth.py:189]: I don't know if using `"ascii"` her is correct?

Comment (clokep) [reply synapse/api/auth.py:189]: Should we have a `hs.get_user_agent_from_request` method like `get_ip_from_request` to consolidate all this logic into one spot at least?

Comment (erikjohnston) [reply synapse/api/auth.py:189]: Good point well made

File: synapse/handlers/auth.py [hunk 2]
```diff
@@ -470,9 +470,7 @@ async def check_ui_auth(
             # authentication flow.
             await self.store.set_ui_auth_clientdict(sid, clientdict)
 
-        user_agent = request.requestHeaders.getRawHeaders(b"User-Agent", default=[b""])[
-            0
-        ].decode("ascii", "surrogateescape")
+        user_agent = request.get_user_agent(" ")
```
Comment (clokep) [inline synapse/handlers/auth.py:473]: I think you want this to be an empty string by default?

```suggestion
        user_agent = request.get_user_agent("")
```

**Review by clokep — APPROVED** [review]

File: synapse/handlers/auth.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/handlers/auth.py]: :facepalm:

**Comment by clokep** [comment]
Looks like the OpenID tests are failing due to a mock missing some parameters.

**Review by clokep — APPROVED** [review]
Still looks good!

**Pull request merged**